### ISSUE TITLE
Questions about customizing training procedures

### ISSUE THREAD

**Issue by mengna0707** [larq/larq#720]
### Describe the bug

I am trying to use a custom training procedure (inherit keras.model class, redefine a class, train_step and test_step are both defined), using this method to train the quantization network defined by larq, there is a strange situation. When kernels were quantified, the loss value in training was very large. It was normal when activation was quantified only. However, when kernel and activation were quantified simultaneously, the value of training Loss seemed to be within the normal range, but loss did not decrease and accuracy did not increase. But everything looked perfect when I trained directly with model.fit(). 
### To Reproduce
```python
class CNN_model(keras.Model):
    def __init__(self, conv_model,  compiled_acc):
        super(CNN_model, self).__init__()
        self.conv_model = conv_model
        self.compiled_acc = compiled_acc

    def train_step(self, data):
        x, y = data
        softmax = keras.layers.Softmax()
        with tf.GradientTape() as tape:
            logits = self.conv_model(x)
            loss_value = self.compiled_loss(y, softmax(logits))
            acc = self.compiled_acc(y,softmax(logits))  # acc
        grads = tape.gradient(loss_value, self.conv_model.trainable_variables) 
        self.optimizer.apply_gradients(zip(grads, self.conv_model.trainable_variables))  
        self.compiled_metrics.update_state(y, softmax(logits))
        return {'train_loss': loss_value, "train_acc":acc}
    
    def test_step(self, data):
        x, y = data
        softmax = keras.layers.Softmax()
        logits = self.conv_model(x)
        loss_value = self.compiled_loss(y, softmax(logits))
        val_acc = self.compiled_acc(y, softmax(logits))  
        return {'loss':loss_value, "val_acc":val_acc}

    def call(self, inputs):
        return self.conv_model(inputs)
 ```
### Expected behavior

I think the custom training process works just as well as using model.fit directly.
### Environment

TensorFlow version: 2.4.0
Larq version: 0.12.0

**Comment by jneeven** [comment]
Hi! This sounds like a bug, I don't think there is conceptually any reason why this shouldn't work. Unfortunately I won't have time to look into it right away, but it'd be very helpful if you could extend the code snippet above to include some model and dataset (don't need to be the ones you're currently using; just whatever model and dataset you can find to reproduce this bug) and the non-subclassed counterpart that you mention does work fine. I'll then be sure to check if there is any bug in Larq that might cause this behaviour...

**Comment by lgeiger** [comment]
As Jelmer, mentioned above it would be great to have a minimal reproducible example to be able to debug this.

Just to verify that the custom training loop is working as expected, could you try setting changing:
`logits = self.conv_model(x)` to `logits = self.conv_model(x, training=True)`?

**Comment by mengna0707** [comment]
Hi @jneeven ,thanks your advice, when I extended the code snippet，I tried the suggestion that  @lgeiger guided，I found my problems solved. So amazing！Thank you @lgeiger  very much ！

But I have another question：when I defined the neural network using tensorflow.keras. layers，everything seemed  normal. When I use larq，I need to change the setting：  logits = self.conv_model(x)  to logits = self.conv_model(x, training=True). I haven't found much explanation for this. Could you please give me some answers

**Comment by simonmaurer** [comment]
@mengna0707 hi there. If your model contains layers that exhibit different behavior for training than evaluation (whether defined as a subclass or directly using tf.keras.models.Model) you need to pass `training=True`.
The .fit() method you usually call actually triggers the overriden method train_step() under the hood.
Check out:
https://github.com/keras-team/keras/blob/3a33d53ea4aca312c5ad650b4883d9bac608a32e/keras/engine/training.py#L787

**Comment by mengna0707** [comment]
@simonmaurer Oh，thank you very much for letting me know this important point that I never thought of.
I think I have a better understanding of tf.keras.models.Model.
Thank you again for your answer！

**Comment by lgeiger** [comment]
Sounds great! Glad this fixes it for you.
